[QHC-1193] make electrical delay a parameter in the drivers.

elygoner opened the pull request.

elygoner committed:
added electrical delay, redy to be tested

elygoner committed:
It was simpler than I thought

elygoner committed:
Merge branch 'main' into qhc-1193-make-electrical-delay-an-actual-physical-parameter-in-the

elygoner committed:
to make mypy shut up

elygoner committed:
mypy dont you dare

DZPM reviewed (approved):
LGTM

elygoner committed:
David's suggestion

flavie-lebars commented on src/qililab/instruments/keysight/e5080b_vna.py:
```diff
@@ -445,7 +445,11 @@ def set_parameter(
+                raise ValueError("Electrical delay has to be between -1e10 and 1e10 nanoseconds")
             self.settings.electrical_delay = float(value)
+            if self.is_device_active():
+                self.device.electrical_delay(self.electrical_delay / 1e9)  # type: ignore [operator]
```
we tend to avoid adding mypy type ignore comment (though sometimes it cannot really be avoided).
Here you can avoid it by checking that it is not None: `if self.is_device_active() and self.electrical_delay is not None:`

flavie-lebars reviewed (commented):
Great work 😊!!
I have left some comments.
Also, please add a changelog.
And the vna driver is already in qcodes so you will need to do a pull request on qcodes to integrate the change of the electrical delay, once this PR has been merged. I can help with this.

elygoner committed:
nanoseconds :P

flavie-lebars reviewed (changes requested):
Adding the changelog as a request here as well so we don't merge by mistake

flavie-lebars committed:
qhc-1216-bug-qblox-draw-should-handle-variable-offsets (#1049)

QbloxDraw: Variable offsets can now be plotted.

elygoner committed:
change log

elygoner committed:
the catual command

elygoner committed:
to nanoseconds

elygoner committed:
Merge branch 'main' into qhc-1193-make-electrical-delay-an-actual-physical-parameter-in-the

elygoner committed:
fixed small issue

flavie-lebars commented on docs/releases/changelog-dev.md:
```diff
@@ -1,6 +1,8 @@
 # Release dev (development release)
 
 ### New features since last release
+- Updated `electrical_delay` to be a changable parameter in the phisical device and not just a software setting used in the autoploting.
```
typos: physical, changeable, and plotting
and please specify the instrument

elygoner committed:
Merge branch 'main' into qhc-1193-make-electrical-delay-an-actual-physical-parameter-in-the

flavie-lebars commented on src/qililab/instruments/keysight/driver_keysight_e5080b.py:
```diff
@@ -309,6 +309,18 @@ def __init__(self, name: str, address: str, **kwargs: Any) -> None:
         )
         """Status Operation"""
 
+        # The more intuitive unit for this parameter is nanosecoonds, even the instrument documentation uses nanoseconds to exemplify, that's why this unit was chosen.
```
typo nanonseconds

flavie-lebars commented on tests/instruments/keysight/test_e5080b.py:
```diff
@@ -280,14 +281,10 @@ def test_get_parameter_method(
+        raw = expected_value.value if isinstance(expected_value, Enum) else expected_value
+        getattr(e5080b_get_param.device, attr_map[parameter_get]).get.return_value = raw
+
+        value = e5080b_get_param.get_parameter(parameter=parameter_get, channel_id=2)
```
Why is there a channel_id? This does not apply to the vna.
This is out of the scope of the PR, we need to improve this validation process.

flavie-lebars commented on tests/instruments/keysight/test_e5080b.py:
```diff
@@ -280,14 +281,10 @@ def test_get_parameter_method(
-            getattr(e5080b_get_param.device, attr_map[parameter_get]).get.return_value = raw
-
-        value = e5080b_get_param.get_parameter(parameter=parameter_get)
+        raw = expected_value.value if isinstance(expected_value, Enum) else expected_value
```
duplicated line

flavie-lebars commented on src/qililab/instruments/keysight/driver_keysight_e5080b.py:
```diff
@@ -307,6 +307,16 @@ def __init__(self, name: str, address: str, **kwargs: Any) -> None:
         )
         """Status Operation"""
 
+        self.electrical_delay: Parameter = self.add_parameter(
```
why was the comment marked as resolved?

elygoner replied:
I don't know why it was marked as resolved but now it is.

flavie-lebars reviewed (changes requested):
1. Have you created a task to add the changes made to the driver to qcodes?
2. electrical delay is missing from the `update_settings` function in e5080b_vna file.

Please redo the tests following https://www.notion.so/qilimanjaro/Testing-procedure-when-a-new-parameter-is-added-to-the-wrapper-or-driver-2037eec14c538023914af3748e23cd83?source=copy_link
the update_settings would have been caught there.

elygoner committed:
add seconds

elygoner committed:
suggested changes

elygoner committed:
update settings

elygoner committed:
added get_parameter if no device connected

elygoner committed:
Ruff

elygoner committed:
Merge branch 'main' into qhc-1193-make-electrical-delay-an-actual-physical-parameter-in-the

elygoner committed:
duplicated test name

elygoner committed:
changelog update

flavie-lebars commented:
Have you addressed those comments @elygoner ?
https://github.com/qilimanjaro-tech/qililab/pull/1047#pullrequestreview-4466735353

elygoner commented:
> Have you addressed those comments @elygoner ? [#1047 (review)](https://github.com/qilimanjaro-tech/qililab/pull/1047#pullrequestreview-4466735353)

I believe I have.

elygoner committed:
Merge branch 'main' into qhc-1193-make-electrical-delay-an-actual-physical-parameter-in-the

flavie-lebars reviewed (approved):
Great work Ely!
